src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -498,14 +500,55 @@ extern "C" void PalAttachThread(void* thread)
 }
 
 #ifdef TARGET_LINUX
-int __cxa_thread_atexit(void (*func)(void*), void* t, void* c)
+struct Link
+{
+    Link* Next;
+    Destructor Dtor;
+    void* Arg;
+};
+
+static pthread_key_t key;
+static pthread_once_t once = PTHREAD_ONCE_INIT;
+
+static void CallDestructors(void* obj)
+{
+    Link* next;
+    Link* link = (Link*)obj;
+    while (link)
+    {
+        next = link->Next;
+        link->Dtor(link->Arg);
+        free(link);
+        link = next;
+    }
+}
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
+}
+
+int __cxa_thread_atexit(Destructor dtor, void* arg, void* dso_symbol)
 {
     if (g_cxa_thread_atexit_impl)
     {
-        return g_cxa_thread_atexit_impl(func, t, c);
+        return g_cxa_thread_atexit_impl(dtor, arg, dso_symbol);
+    }
+
+    // fallback implementation
+
+    Link *link;
+    pthread_once(&once, CreateKey);
+    link = (Link*)malloc(sizeof(Link));
+    if (link)
+    {
+        link->Next = (Link*)pthread_getspecific(key);
```

Comment:
A few related observations (repo-wide):

* When C++ compiler encounters `thread_local` keyword on linux, it expects `__cxa_thread_atexit` symbol to be present (provided by libstdc++).
* C11 has an alternative `_THREAD_local`, but msvc [does not support it yet](https://developercommunity.visualstudio.com/t/C-compiler:-_Thread_local-is-not-suppore/10020499?space=62&q=_Thread_local)
  * cross-dac uses the empty declaration in hp-libunwind on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/native/external/libunwind_extras/configure.cmake#L22-L25
  * mono uses declspec(thread)  in C on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/mono/CMakeLists.txt#L272
  * nativeaot also uses declspec(thread) on windows and __thread on unix https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/coreclr/nativeaot/Runtime/CommonMacros.h#L235-L238 note that __thread (gcc/clang extension) and _THREAD_LOCAL (C11) on unix have the same semantics.
  * llvm-libunwind also uses declspec(thread) in setjump/longjump sjlj on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/native/external/llvm-libunwind/src/Unwind-sjlj.c#L68 but we do not compile this code in nativeaot (we only compile the alternative: Unwind-EHABI.cpp / dwarf as sjlj is typically used by mingw based toolchain which we don't support)

Since C11 and C++11 are minimal requirements, we can safely use `_THREAD_LOCAL` on unix and `declspec(thread)` on windows everywhere, in order to get rid of `__thread` (gcc extension) and `thread_local` (C++11 keyword) variants.

Reply:
> we can safely use _THREAD_LOCAL on unix and declspec(thread) on windows everywhere,

This sounds good to me for the minimal native AOT runtime.

CoreCLR can stay with `thread_local` since it has heavy dependency on C++ runtime. No need to change it there.

Reply:
> CoreCLR can stay with `thread_local` since it has heavy dependency on C++ runtime. No need to change it there.

Sure. Btw, CoreCLR is using both, `thread_local` (more often) and `declspec(thread)` (in a few places).

Reply:
I have made the fallback path to be taken on Unix (both Linux and macOS). To use C11's `_Thread_local` and `tss_create` etc. we would need to convert the source to `.c` to be able to use `#include <threads.h>`, as C++11's `#include <thread>` requires std::new, which then conflicts with our shim.

I will try to give a separate PR pass to convert NativeAOT sources to C (there are quite many changes to make it happen:; new->malloc, delete->free, class->struct etc.). It will also require discussions with GC team; mainly to find out if C++ to C conversion is a viable option for their development.

Reply:
> requires std::new, which then conflicts with our shim.

Do you mean no-throw new overload? Can we create our own no-throw new overload to avoid conflicts?

I think converting GC to C would be too much of a stretch at this point. I think sticking with C-with-classes is much more workable.

Reply:
Sounds like we need to port the GC to C# 😄

Reply:
> Can we create our own no-throw new overload to avoid conflicts?

We have no-throw overload in `stdcppshim.cpp`. If we use `#include <thread>` (or `#include <algorithm>` etc.), those headers in turn use `#include <new>` (which requires libstdc++ runtime library) and conflict with our shim.

Reply:
> Sounds like we need to port the GC to C# 😄

There is a hope: https://github.com/dotnet/runtime/discussions/72037 🙂